pkg/controller/vitesscell/reconcile_vtgate.go

```diff
@@ -170,5 +175,35 @@ func (r *ReconcileVitessCell) reconcileVtgate(ctx context.Context, vtc *planetsc
 		resultBuilder.Error(err)
 	}
 
+	var wantHpa bool
+	var hpaSpec *vtgate.HpaSpec
+
+	if vtc.Spec.Gateway.Autoscaler != nil {
+		wantHpa = vtc.Spec.Gateway.Autoscaler.MaxReplicas != nil
+		hpaSpec = &vtgate.HpaSpec{
+			Labels:      labels,
+			MinReplicas: vtc.Spec.Gateway.Autoscaler.MinReplicas,
+			MaxReplicas: vtc.Spec.Gateway.Autoscaler.MaxReplicas,
+			Behavior:    vtc.Spec.Gateway.Autoscaler.Behavior,
+			Metrics:     vtc.Spec.Gateway.Autoscaler.Metrics,
+		}
+	}
+
+	// Reconcile vtgate HorizontalPodAutoscaler.
+	err = r.reconciler.ReconcileObject(ctx, vtc, key, labels, wantHpa, reconciler.Strategy{
+		Kind: &autoscalingv2.HorizontalPodAutoscaler{},
+
+		New: func(key client.ObjectKey) runtime.Object {
+			return vtgate.NewHorizontalPodAutoscaler(key, hpaSpec)
+		},
+		UpdateInPlace: func(key client.ObjectKey, obj runtime.Object) {
+			newObj := obj.(*autoscalingv2.HorizontalPodAutoscaler)
+			vtgate.UpdateHorizontalPodAutoscaler(newObj, hpaSpec)
+		},
+	})
+	if err != nil {
+		resultBuilder.Error(err)
+	}
```

Comment:
Without this change, creating a cluster does not fail as the vitess-operator can still run. But no HPA is created:
```
$ kubectl get hpa
No resources found in default namespace.
```

With this change however, it looks good, vitess-operator is healthy and we can see the HPA:
```
$ kubectl get hpa
NAME                            REFERENCE                                  TARGETS         MINPODS   MAXPODS   REPLICAS   AGE
example-zone1-vtgate-bc6cde92   VitessCell/example-zone1-vtgate-bc6cde92   <unknown>/10%   1         10        0          33s
```

Note that the target is unknown because my `metrics-server` was disabled at the time.

Reply:
I added the metrics-server in the [test/endtoend/operator/101_initial_cluster_autoscale.yaml](https://github.com/siadat/vitess-operator/blob/sina/vitesscell-hpa/test/endtoend/operator/101_initial_cluster_autoscale.yaml#L207-L409) and used it in the hpa endtoend test.

This way I can verify that the specs are correct because the target object is actually auto-scaled from 1 replicas to the minimum 2 replicas.

Reply:
This is a great idea, thank you 🙏🏻